src/Sylius/Bundle/ThemeBundle/Templating/Locator/TemplateLocator.php

```diff
@@ -0,0 +1,270 @@
+            }
+        }
+
+        if (0 === strpos($template->getPath(), '@')) {
+            return $this->locateBundleTemplateUsingThemes($template, $themes);
+        } else {
+            return $this->locateAppTemplateUsingThemes($template, $themes);
+        }
+    }
+
+    /**
+     * @param TemplateReferenceInterface $template
+     * @param ThemeInterface[] $themes
+     *
+     * @return null|string
+     */
+    protected function locateBundleTemplateUsingThemes(TemplateReferenceInterface $template, array $themes)
+    {
+        $name = $template->getPath();
+
+        if (false !== strpos($name, '..')) {
+            throw new \RuntimeException(sprintf('File name "%s" contains invalid characters (..).', $name));
+        }
+
+        $bundleName = substr($name, 1);
+        $templatePath = '';
+        if (false !== strpos($bundleName, '/')) {
+            list($bundleName, $templatePath) = explode('/', $bundleName, 2);
+        }
+        if (0 !== strpos($templatePath, 'Resources')) {
+            throw new \RuntimeException('Template files have to be in Resources.');
+        }
+
+        $resourceBundle = null;
+        $bundles = $this->kernel->getBundle($bundleName, false);
+
+        $parameters = [
+            '%app_path%' => $this->appDir,
+            '%override_path%' => substr($templatePath, strlen('Resources/')),
+        ];
+
+        foreach ($this->paths['bundle_template'] as $path) {
+            foreach ($bundles as $bundle) {
+                $parameters = array_merge($parameters, [
+                    '%bundle_name%' => $bundle->getName(),
+                    '%bundle_path%' => $bundle->getPath(),
+                ]);
+
+                if (false === strpos($path, '%theme_path%')) {
+                    if (null !== $checkedPath = $this->checkPath($path, $parameters, $template)) {
+                        return $checkedPath;
+                    }
+                } else {
+                    foreach ($themes as $theme) {
+                        $themeParameters = array_merge($parameters, [
+                            '%theme_path%' => $theme->getPath(),
+                        ]);
+
+                        if (null !== $checkedPath = $this->checkPath($path, $themeParameters, $template, $theme)) {
+                            return $checkedPath;
+                        }
+                    }
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param TemplateReferenceInterface $template
+     * @param ThemeInterface[] $themes
+     *
+     * @return null|string
+     */
+    protected function locateAppTemplateUsingThemes(TemplateReferenceInterface $template, array $themes = [])
+    {
+        $parameters = [
+            '%app_path%' => $this->appDir,
+            '%override_path%' => $template->getPath(),
+        ];
+
+        foreach ($this->paths['app_template'] as $path) {
+            if (false === strpos($path, '%theme_path%')) {
```

Comment:
What do you think about this condition?
```php
if (false !== strpos($path, '%theme_path%')) {
    foreach ($themes as $theme) {
        $themeParameters = array_merge($parameters, [
            '%theme_path%' => $theme->getPath(),
        ]);
        if (null !== $checkedPath = $this->checkPath($path, $themeParameters, $template, $theme)) {
            return $checkedPath;
        }
    }
}
if (null !== $checkedPath = $this->checkPath($path, $parameters, $template)) {
    return $checkedPath;
}
```
Futhermore we can extract foreach loop into private method.